File: src/main/java/FunctionTable.java
import java.util.Map;
import java.util.HashMap;

public class FunctionTable {
    private Map<String, Function> table = new HashMap<String, Function>();

    public FunctionTable() {
        set(new PrintFunction());
    }

    public Function get(String name) {
        return table.get(name);
    }

    public void set(Function function) {
        table.put(function.getName(), function);
    }

    private class PrintFunction extends Function {
        public PrintFunction() {
            super("print", null, null);
        }

        @Override
        public Value call(Value ... args) {
            System.out.print(args[0].toString());
            return null;
        }
    }
}

File: src/main/scala/FunctionTable.scala
import java.util.HashMap

class FunctionTable {
    private val table = new HashMap[String, Function]

    set(new PrintFunction())

    def get(name:String):Function = table.get(name)
    def set(function:Function) = table.put(function.name, function)

    class PrintFunction extends Function("print", null, null) {

        override def call(args:Value*):Value = {
            System.out.print(args(0).stringValue)
            null
        }
    }
}
